snapshot/pkg/controller/snapshotter/snapshotter.go

```diff
@@ -382,56 +381,35 @@ func (vs *volumeSnapshotter) getSnapshotStatus(snapshot *crdv1.VolumeSnapshot) (
 	if status == statusReady {
 		return status, nil, nil
 	} else if status == statusPending {
-		// If we are here, takeSnapshot has already happened.
-		// Check whether the VolumeSnapshotData object is already created
+		// If we are here, takeSnapshot has already happened and
+		// VolumeSnapshotData object is already created
 		snapshotDataObj = vs.getSnapshotDataFromSnapshotName(snapshotName)
 		if snapshotDataObj == nil {
-			bCreateSnapData = true
-			// Find snapshot by existing tags, and create VolumeSnapshotData
-		} else {
-			// Bind VolumeSnapshotData to VolumeSnapshot if it has not happened yet
-			if snapshot.Spec.SnapshotDataName != snapshotDataObj.Metadata.Name {
-				glog.Infof("getSnapshotStatus: bind VolumeSnapshotData to VolumeSnapshot %s.", snapshotName)
-				err = vs.bindVolumeSnapshotDataToVolumeSnapshot(snapshotName, snapshotDataObj.Metadata.Name)
-				if err != nil {
-					glog.Errorf("getSnapshotStatus: Error updating volume snapshot %s: %v", snapshotName, err)
-					return statusError, nil, err
-				}
-			}
-			return status, snapshotDataObj, nil
+			return statusError, nil, fmt.Errorf("VolumeSnapshotData not found for snapshot %s", snapshotName)
 		}
+		return status, snapshotDataObj, nil
 	} else if status == statusError {
 		return status, nil, fmt.Errorf("Failed to find snapshot %s", snapshotName)
 	}
-	bCreateSnapData = true
-	// Find snapshot by existing tags, and create VolumeSnapshotData
 
-	if bCreateSnapData {
-		snapshotDataSource, conditions, err = vs.findSnapshot(snapshotName, snapshot)
-		if err != nil {
-			return statusNew, nil, nil
-		}
-		glog.Infof("getSnapshotStatus: create VolumeSnapshotData object for VolumeSnapshot %s.", snapshotName)
-		snapshotDataObj, err := vs.createVolumeSnapshotData(snapshotName, snapshot, snapshotDataSource)
-		if err != nil {
-			return statusError, nil, err
-		}
-		if status != statusReady {
-			_, err = vs.UpdateVolumeSnapshot(snapshotName, conditions)
-			if err != nil {
-				glog.Errorf("getSnapshotStatus: Error updating volume snapshot %s: %v", snapshotName, err)
-				return statusError, nil, err
-			}
-		}
-		glog.Infof("getSnapshotStatus: bind VolumeSnapshotData to VolumeSnapshot %s.", snapshotName)
-		err = vs.bindVolumeSnapshotDataToVolumeSnapshot(snapshotName, snapshotDataObj.Metadata.Name)
-		if err != nil {
-			glog.Errorf("getSnapshotStatus: Error binding VolumeSnapshotData to VolumeSnapshot %s: %v", snapshotName, err)
-			return statusError, nil, err
-		}
-		return statusPending, snapshotDataObj, nil
+	// Find snapshot by existing tags, and create VolumeSnapshotData
```

Comment:
Shouldn't we split the function here? It's named 'getSnapshotStatus', so it's rather surprising that is also creates VolumeSnapshotData and binds them to the VolumeSnapshot. It's also called just from syncSnapshot so perhaps returning here and calling another method that creates and binds VolumeSnapshotData from syncSnapshot when needed would improve the readability.

Reply:
The function name "getSnapshotStatus" is confusing.  Maybe rename it to "checkAndUpdateSnapshot"?

Reply:
I don't see that as an improvement: The function does not update the status, it creates and binds VolumeSnapshotData based on the status. Splitting it in two functions (one to get the status and other to create and bind the object) looks more "natural" to me.

Reply:
Do you mean to add crdv1.VolumeSnapshotConditionPendingMissingData as well so CloudProvider will only return this new state, or do you mean only add statusPendingMissingData within snapshotter.go itself (so we'll have statusPending plus statusPendingMissingData)?

Reply:
I split out the code to find snapshot, create VolumeSnapshotData, and update VolumeSnapshot status and bind VolumeSnapshotData to VolumeSnapshot to a new function called updateSnapshotIfExists.